drivers/sensor/pmw3360/pmw3360.c

```diff
@@ -0,0 +1,947 @@
+	err = reg_write(dev_data, PMW3360_REG_SROM_ENABLE, 0x18);
+	if (err) {
+		LOG_ERR("Cannot start SROM download");
+		return err;
+	}
+
+	/* Write SROM file into SROM_Load_Burst register.
+	 * Data must start with SROM_Load_Burst address.
+	 */
+	err = burst_write(dev_data, PMW3360_REG_SROM_LOAD_BURST,
+			  pmw3360_firmware_data, pmw3360_firmware_length);
+	if (err) {
+		LOG_ERR("Cannot write firmware to sensor");
+	}
+
+	return err;
+}
+
+static int pmw3360_async_init_fw_load_verify(struct pmw3360_data *dev_data)
+{
+	int err;
+
+	/* Read the SROM_ID register to verify the firmware ID before any
+	 * other register reads or writes
+	 */
+
+	u8_t fw_id;
+	err = reg_read(dev_data, PMW3360_REG_SROM_ID, &fw_id);
+	if (err) {
+		LOG_ERR("Cannot obtain firmware id");
+		return err;
+	}
+
+	LOG_DBG("Optical chip firmware ID: 0x%x", fw_id);
+	if (fw_id != PMW3360_FIRMWARE_ID) {
+		LOG_ERR("Chip is not running from SROM!");
+		return -EIO;
+	}
+
+	u8_t product_id;
+	err = reg_read(dev_data, PMW3360_REG_PRODUCT_ID, &product_id);
+	if (err) {
+		LOG_ERR("Cannot obtain product id");
+		return err;
+	}
+
+	if (product_id != PMW3360_PRODUCT_ID) {
+		LOG_ERR("Invalid product id!");
+		return -EIO;
+	}
+
+	/* Write 0x20 to Config2 register for wireless mouse design.
+	 * This enables entering rest modes.
+	 */
+	err = reg_write(dev_data, PMW3360_REG_CONFIG2, 0x20);
+	if (err) {
+		LOG_ERR("Cannot enable REST modes");
+	}
+
+	return err;
+}
+
+static void irq_handler(struct device *gpiob, struct gpio_callback *cb,
+			u32_t pins)
+{
+	sensor_trigger_handler_t handler;
+
+	k_spinlock_key_t key = k_spin_lock(&pmw3360_data.lock);
+	handler = pmw3360_data.data_ready_handler;
+	k_spin_unlock(&pmw3360_data.lock, key);
+
+	struct sensor_trigger trig = {
+		.type = SENSOR_TRIG_DATA_READY,
+		.chan = SENSOR_CHAN_ALL,
+	};
+	handler(DEVICE_GET(pmw3360), &trig);
+}
+
+static int pmw3360_async_init_power_up(struct pmw3360_data *dev_data)
+{
+	/* Reset sensor */
+
+	return reg_write(dev_data, PMW3360_REG_POWER_UP_RESET, 0x5A);
+}
+
+static int pmw3360_async_init_configure(struct pmw3360_data *dev_data)
+{
+	int err;
+
+	err = update_cpi(dev_data, CONFIG_PMW3360_CPI);
+
+	if (!err) {
+		err = update_downshift_time(dev_data,
+					    PMW3360_REG_RUN_DOWNSHIFT,
+					    CONFIG_PMW3360_RUN_DOWNSHIFT_TIME_MS);
+	}
+
+	if (!err) {
+		err = update_downshift_time(dev_data,
+					    PMW3360_REG_REST1_DOWNSHIFT,
+					    CONFIG_PMW3360_REST1_DOWNSHIFT_TIME_MS);
+	}
+
+	if (!err) {
+		err = update_downshift_time(dev_data,
+					    PMW3360_REG_REST2_DOWNSHIFT,
+					    CONFIG_PMW3360_REST2_DOWNSHIFT_TIME_MS);
+	}
+
+	if (!err) {
+		dev_data->ready = true;
+	}
+
+	return err;
+}
+
+static void pmw3360_async_init(struct k_work *work)
+{
+	struct pmw3360_data *dev_data = &pmw3360_data;
+
+	ARG_UNUSED(work);
+
+	LOG_DBG("PMW3360 async init step %d", dev_data->async_init_step);
+
+	dev_data->err = async_init_fn[dev_data->async_init_step](dev_data);
+	if (dev_data->err) {
```

Comment:
Does the `err` need to be a field of `dev_data`? It does not seem to be used anywhere else in the code. Is it kept for debugging purposes?

Reply:
Debugging. I though I would work to find a way for app to sync to initialization completion.
The device API as such gives little possibility. 

- We could use a power management call to set driver to active state. This would block until device is initialize and in case there was an error a known error would be returned. But this will not work when PM is disabled in the system.

- Better way would be to extend the device API with a new call that would allow application to block until a device is ready. I though I could do something with `device_busy_check` infrastructure but this keeps it's data in the special section and would required adding some pollable object that app could hang on. I would prefer it. At the moment however I don't have time to dig into implementing this especially that it would be pain to get it into zephyr quickly.

I probably get back to it later on while working on the second sensor.